File: chime/extractors/person_aliases.py
"""Familiar public names for well-known CSE directors listed by initials.

CSE ``companyProfile`` often stores only initials (``K. A. D. D. Perera``).
These aliases are for display / soft-merge only — ``name_norm`` stays the
CSE initials form as the stable identity key.

Only high-confidence, widely published mappings. Do not invent names.
"""

from __future__ import annotations

# name_norm → (canonical_merge_key, preferred_display_name)
PERSON_ALIASES: dict[str, tuple[str, str]] = {
    # Vallibel / Hayleys
    "K A D D PERERA": ("dhammika_perera", "Dhammika Perera"),
    # Hayleys — CSE uses M. on HAYL and A.M. on group companies
    "M PANDITHAGE": ("mohan_pandithage", "Mohan Pandithage"),
    "A M PANDITHAGE": ("mohan_pandithage", "Mohan Pandithage"),
    # John Keells
    "K BALENDRA": ("krishan_balendra", "Krishan Balendra"),
    "K N J BALENDRA": ("krishan_balendra", "Krishan Balendra"),
    "J G A COORAY": ("gihan_cooray", "Gihan Cooray"),
    # Distilleries / Melstacorp family
    "D S T JAYAWARDENA": ("harry_jayawardena", "Harry Jayawardena"),
    "DON S T JAYAWARDENA": ("harry_jayawardena", "Harry Jayawardena"),
    "D HASITHA S JAYAWARDENA": ("hasitha_jayawardena", "Hasitha Jayawardena"),
    "D HASITHA STASSEN JAYAWARDENA": (
        "hasitha_jayawardena",
        "Hasitha Jayawardena",
    ),
    # LOLC
    "I C NANAYAKKARA": ("ishara_nanayakkara", "Ishara Nanayakkara"),
    "W D K JAYAWARDENA": ("kapila_jayawardena", "Kapila Jayawardena"),
    # ACL / Lanka Tiles group
    "H A S MADANAYAKE": ("suren_madanayake", "Suren Madanayake"),
    "U G MADANAYAKE": ("udaya_madanayake", "Udaya Madanayake"),
    # Other frequently initials-only chairs
    "S K SHAH": ("suresh_shah", "Suresh Shah"),
    "S H AMARASEKERA": ("harsha_amarasekera", "Harsha Amarasekera"),
    "D A CABRAAL": ("amal_cabraal", "Amal Cabraal"),
}


def preferred_display_name(display_name: str, name_norm: str) -> str:
    key = (name_norm or "").strip().upper()
    hit = PERSON_ALIASES.get(key)
    if hit:
        return hit[1]
    return display_name


def alias_merge_key(name_norm: str) -> str | None:
    key = (name_norm or "").strip().upper()
    hit = PERSON_ALIASES.get(key)
    return hit[0] if hit else None

File: web/src/lib/api/person-aliases.ts
/**
 * Familiar public names for well-known CSE directors listed by initials.
 * Keep in sync with chime/extractors/person_aliases.py
 */

type Alias = { key: string; display: string };

/** Normalized uppercase name (dots → spaces) → alias */
const BY_COMPACT: Record<string, Alias> = {
  "DHAMMIKA PERERA": { key: "dhammika_perera", display: "Dhammika Perera" },
  "K A D D PERERA": { key: "dhammika_perera", display: "Dhammika Perera" },
  "MOHAN PANDITHAGE": { key: "mohan_pandithage", display: "Mohan Pandithage" },
  "M PANDITHAGE": { key: "mohan_pandithage", display: "Mohan Pandithage" },
  "A M PANDITHAGE": { key: "mohan_pandithage", display: "Mohan Pandithage" },
  "KRISHAN BALENDRA": { key: "krishan_balendra", display: "Krishan Balendra" },
  "K BALENDRA": { key: "krishan_balendra", display: "Krishan Balendra" },
  "K N J BALENDRA": { key: "krishan_balendra", display: "Krishan Balendra" },
  "GIHAN COORAY": { key: "gihan_cooray", display: "Gihan Cooray" },
  "J G A COORAY": { key: "gihan_cooray", display: "Gihan Cooray" },
  "HARRY JAYAWARDENA": { key: "harry_jayawardena", display: "Harry Jayawardena" },
  "D S T JAYAWARDENA": { key: "harry_jayawardena", display: "Harry Jayawardena" },
  "DON S T JAYAWARDENA": { key: "harry_jayawardena", display: "Harry Jayawardena" },
  "HASITHA JAYAWARDENA": {
    key: "hasitha_jayawardena",
    display: "Hasitha Jayawardena",
  },
  "D HASITHA S JAYAWARDENA": {
    key: "hasitha_jayawardena",
    display: "Hasitha Jayawardena",
  },
  "D HASITHA STASSEN JAYAWARDENA": {
    key: "hasitha_jayawardena",
    display: "Hasitha Jayawardena",
  },
  "ISHARA NANAYAKKARA": {
    key: "ishara_nanayakkara",
    display: "Ishara Nanayakkara",
  },
  "I C NANAYAKKARA": { key: "ishara_nanayakkara", display: "Ishara Nanayakkara" },
  "KAPILA JAYAWARDENA": {
    key: "kapila_jayawardena",
    display: "Kapila Jayawardena",
  },
  "W D K JAYAWARDENA": {
    key: "kapila_jayawardena",
    display: "Kapila Jayawardena",
  },
  "SUREN MADANAYAKE": { key: "suren_madanayake", display: "Suren Madanayake" },
  "H A S MADANAYAKE": { key: "suren_madanayake", display: "Suren Madanayake" },
  "UDAYA MADANAYAKE": { key: "udaya_madanayake", display: "Udaya Madanayake" },
  "U G MADANAYAKE": { key: "udaya_madanayake", display: "Udaya Madanayake" },
  "SURESH SHAH": { key: "suresh_shah", display: "Suresh Shah" },
  "S K SHAH": { key: "suresh_shah", display: "Suresh Shah" },
  "HARSHA AMARASEKERA": {
    key: "harsha_amarasekera",
    display: "Harsha Amarasekera",
  },
  "S H AMARASEKERA": { key: "harsha_amarasekera", display: "Harsha Amarasekera" },
  "AMAL CABRAAL": { key: "amal_cabraal", display: "Amal Cabraal" },
  "D A CABRAAL": { key: "amal_cabraal", display: "Amal Cabraal" },
};

export function compactPersonName(name: string): string {
  return name
    .replace(/\./g, " ")
    .replace(/\s+/g, " ")
    .trim()
    .toUpperCase();
}

export function personAlias(name: string): Alias | null {
  const compact = compactPersonName(name);
  return BY_COMPACT[compact] ?? null;
}

export function preferredDisplayName(name: string): string {
  return personAlias(name)?.display ?? name;
}

/** Soft-merge key — aliases collapse split CSE spellings. */
export function softPersonKey(name: string): string {
  const alias = personAlias(name);
  if (alias) return `ALIAS:${alias.key.toUpperCase()}`;
  const compact = compactPersonName(name);
  const parts = compact.split(/\s+/).filter(Boolean);
  if (parts.length === 0) return compact;
  const last = parts[parts.length - 1];
  const initials = parts
    .slice(0, -1)
    .map((p) => p[0] ?? "")
    .join("");
  return `${initials}:${last}`;
}
